ipcc/src/lib.rs

```diff
@@ -135,10 +137,84 @@ impl InstallinatorImageId {
     }
 }
 
-// TODO Add ioctl wrappers? `installinator` is the only client for
-// `Key::InstallinatorImageId`, but we might grow other keys for other clients,
-// at which point we probably want all the ioctl wrapping to happen in one
-// place.
+#[derive(Debug, Error)]
+pub enum InstallinatorImageIdError {
+    #[error(transparent)]
+    Ipcc(#[from] IpccError),
+    #[error("deserializing installinator image ID failed: {0}")]
+    DeserializationFailed(String),
+}
+
+#[derive(Error, Debug)]
+pub enum IpccError {
+    #[error("Memory allocation error")]
+    NoMem(#[source] IpccErrorInner),
+    #[error("Invalid parameter")]
+    InvalidParam(#[source] IpccErrorInner),
+    #[error("Internal error occurred")]
+    Internal(#[source] IpccErrorInner),
+    #[error("Requested lookup key was not known to the SP")]
+    KeyUnknown(#[source] IpccErrorInner),
+    #[error("Value for the requested lookup key was too large for the supplied buffer")]
+    KeyBufTooSmall(#[source] IpccErrorInner),
+    #[error("Attempted to write to read-only key")]
+    KeyReadonly(#[source] IpccErrorInner),
+    #[error("Attempted write to key failed because the value is too long")]
+    KeyValTooLong(#[source] IpccErrorInner),
+    #[error("Compression or decompression failed")]
+    KeyZerr(#[source] IpccErrorInner),
+    #[error("Unknown libipcc error")]
+    UnknownErr(#[source] IpccErrorInner),
+}
+
+#[derive(Error, Debug)]
+#[error("{context}: {errmsg} ({syserr})")]
+pub struct IpccErrorInner {
+    pub context: String,
+    pub errmsg: String,
+    pub syserr: String,
+}
+
+/// These are the IPCC keys we can look up.
+/// NB: These keys match the definitions found in libipcc (RFD 316) and should
+/// match the values in `[ipcc::Key]` one-to-one.
+#[derive(Debug, Clone, Copy)]
+#[allow(dead_code)]
+#[repr(u8)]
+enum IpccKey {
+    Ping = 0,
+    InstallinatorImageId = 1,
+    Inventory = 2,
+    System = 3,
+    Dtrace = 4,
+}
+
+/// Interface to the inter-processor communications channel.
+/// For more information see rfd 316.
```

Comment:
are there docs for the illumos library as well?

Reply:
There is no documentation for the library other than the header file.  I thought about linking the ipcc driver man page but the point of libipcc is that users don't call the kernel driver interface directly anymore.

Reply:
There will be man pages for libipcc in the next week or so.